flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java

```diff
@@ -170,9 +173,14 @@
 
 	private final CompletableFuture<Void> closeFuture;
 
+	@Nullable
+	protected BufferReceivedListener bufferReceivedListener;
```

Comment:
+1 for trying to make it `final`. As it is now, the circular dependency is a bit fishy, but I'm not entirely sure how can this be dealt with?

I think currently there are no concurrency issues, as this field is being set very shortly after the construction, but that's a bit fragile.

Reply:
Regarding 
> I'd refrain from using volatile on the hot path

I share this concern but
> we could cache volatile value if we guarantee that it's never changed once set

Reply:
In general, I also hate the non-final fields in class. But in this class, it seems not easy to make it final, because the `SingleInputGate` is created by `ShuffleEnvironment` in task constructor, which belongs to shuffle service component. And `CheckpointedInputGate`, `UnionInputGate`, `CheckpointBarrierHandler` are constructed inside task invokable, which belongs to task component.  So we can not get the `CheckpointBarrierHandler` instance while constructing `SingleInputGate`.

Another way to handle this case is via monitoring the future from serious upper to lower components, just like the existing way that `UnionInputGate` monitors data available from `SingleInputGate`. That means when the `SingleInputGate` receives barrier or buffer from network, it completes some future to make it aware by upper `UnionInputGate`, `CheckpointedInputGate`.  But it seems more complex with existing data available future.

So I agree with the option of volatile and guarantee it is set only once.

Reply:
Yes added a guard to avoid it being reset.

Reply:
Just adding a guard is not enough to protect against visibility issues.

To clarify, I don't agree with @pnowojski:
> I think currently there are no concurrency issues, as this field is being set very shortly after the construction, before passing it to different threads?

IMO, it doesn't matter whether it is set during construction or not if there are no memory barriers.

And since `final` is not an option we are only left with `volatile`. Which as I said we can be cached. 

By caching I mean:
1. mark the field `volatile`
1. "constructing thread" writes the value before any following reads
1. "reading thread" uses a non-volatile field which is updated once per thread in the beginning:
```
if (cachedBufferReceivedListener == null) {
  cachedBufferReceivedListener = bufferReceivedListener;
}
cachedBufferReceivedListener.notify(...)
```

I'm not sure whether it's better to place `cachedBufferReceivedListener` in the input channel or input gate.

(please unresolve this thread)

Reply:
This approach is an elegant solution in the case of unaligned checkpoints. However, on aligned checkpoint, it adds a volatile access to each access (as it's currently never non-null). 
I could make the aligned case use a no-op listener, but that would cause unnecessary buffer copying in `RemoteInputChannel`...

Reply:
Okay I opted to go with a volatile field now, but it's properly encapsulated in a getter for future optimizations. I'd start simple first and we can always make it more complicated after benchmarks demand it.
I'm hoping that a `volatile` read access right before entering synchronized would kind of synergize, such that the cache pipe of the CPU is only cleared once.

Reply:
I don't think `synchronized` will speed up `volatile` read. But I'm also not sure what the performance impact is in the first place so I'm fine with this approach.